cogs/boolean_utils.py

```diff
@@ -0,0 +1,395 @@
+                # Compute result and add it back to stack
+                if char == '!':
+                    one = not post_stack.pop()
+                    post_stack.append(one)
+                if char == '*':
+                    one = post_stack.pop()
+                    two = post_stack.pop()
+                    result = one and two
+                    post_stack.append(result)
+                if char == '+':
+                    one = post_stack.pop()
+                    two = post_stack.pop()
+                    result = one or two
+                    post_stack.append(result)
+                if char == '^':
+                    one = post_stack.pop()
+                    two = post_stack.pop()
+                    result = one and not two or two and not one
+                    post_stack.append(result)
+        # Save result in a formatted fashion, will be single line in table
+        self.format_result(post_stack.pop(), dict_bool, vars)
+
+    def format_header(self, vars):
+        '''Creates the header to the truth table'''
+        self.result_formatted += "+" + "---+" * len(vars)
+        self.result_formatted += "-" * len(self.disp_exp) + "--+\n"
+        for var in vars:
+            self.result_formatted += "| {} ".format(var)
+        self.result_formatted += "| {} |\n".format(self.disp_exp)
+        self.result_formatted += "+---" * len(vars)
+        self.result_formatted += "+" + "-" * len(self.disp_exp) + "--+\n"
+
+    def format_result(self, result, dict_bool, vars):
+        '''Creates single line in the truth table'''
+        for var in vars:
+            booly = self.bool_to_int(dict_bool[var])
+            self.result_formatted += "| {} ".format(booly)
+        res = self.bool_to_int(result)
+        self.result_formatted += "| {}".format(res)
+        self.result_formatted += " " * len(self.disp_exp) + "|\n"
+
+    def bool_to_int(self, booly):
+        '''Returns 1 if True, 0 if False'''
+        if booly == True:
+            return 1
+        else:
+            return 0
+
+    def valid_vars(self):
+        '''False if there are two vars right next to each other'''
+        for i in range(1, len(self.orig_exp)):
+            if self.orig_exp[i].isalpha() and self.orig_exp[i - 1].isalpha():
+                return False
+        return True
+
+    def valid_negation(self):
+        '''False if operator! is placed incorrectly'''
+        for i in range(1, len(self.orig_exp)):
+            if self.orig_exp[i] == '!' and self.orig_exp[i - 1].isalpha():
+                self.error_msg = "! should come before a variable, not after."
+                return False
+            if self.orig_exp[i] == '!' and self.orig_exp[i - 1] == ')':
+                self.error_msg = "'!' should not follow a ')'"
+                return False
+            if self.orig_exp[i] == '!' and self.orig_exp[i - 1] == '!':
+                self.error_msg = "Please remove redundant ! operators..."
```

Comment:
What about doing `str.Replace('!!', '')`, as a easy way to catch stuff like this

Reply:
That sounds like a dandy idea